SourceX/qol.cpp

```diff
@@ -155,39 +190,103 @@ void DrawXPBar(CelOutputBuffer out)
+
+	// Figure out how much to fill the last pixel of the XP bar, to make it gradually appear with gained XP
+	Uint64 onePx = prevXpDelta / xpbar::barWidth;
+	Uint64 lastFullPx = fullBar * prevXpDelta / xpbar::barWidth;
+
+	const Uint64 fade = (prevXpDelta_1 - lastFullPx) * (xpbar::whiteGrades - 1) / onePx;
+
+	// Draw beginning of bar full brightness
+	FastDrawHorizLine(out, xPos, yPos + 1, fullBar, xpbar::whiteGradient[xpbar::whiteGrades * 3 / 4 - 1]);
+	FastDrawHorizLine(out, xPos, yPos + 2, fullBar, xpbar::whiteGradient[xpbar::whiteGrades - 1]);
+	FastDrawHorizLine(out, xPos, yPos + 3, fullBar, xpbar::whiteGradient[xpbar::whiteGrades / 2 - 1]);
+
+	// End pixels appear gradually
+	SetPixel(out, xPos + fullBar, yPos + 1, xpbar::whiteGradient[fade * 3 / 4]);
+	SetPixel(out, xPos + fullBar, yPos + 2, xpbar::whiteGradient[fade]);
+	SetPixel(out, xPos + fullBar, yPos + 3, xpbar::whiteGradient[fade / 2]);
+}
+
+bool CheckXPBarInfo(void)
+{
+	if (!sgOptions.Gameplay.bExperienceBar)
+		return false;
+
+	const int backX = (gnScreenWidth / 2) - 155;
+	const int backY = gnScreenHeight - 11;
+
+	if (MouseX < backX || MouseX >= backX + xpbar::backWidth ||
+		MouseY < backY || MouseY >= backY + xpbar::backHeight)
+		return false;
+
+	const PlayerStruct& player = plr[myplr];
+
+	const int charLevel = player._pLevel;
+
+	char tempstr[64];
+
+	if (charLevel == MAXCHARLEVEL - 1) {
+		// Show a maximum level indicator for max level players.
+		infoclr = COL_GOLD;
+
+		sprintf(tempstr, "Level %d", charLevel);
+		AddPanelString(tempstr, true);
+
+		sprintf(tempstr, "Experience: ");
+		PrintWithSeparator(tempstr + SDL_arraysize("Experience: ") - 1, ExpLvlsTbl[charLevel - 1]);
+		AddPanelString(tempstr, true);
+
+		AddPanelString("Maximum Level", true);
+
+		return true;
+	}
+
+	infoclr = COL_WHITE;
+
+	sprintf(tempstr, "Level %d", charLevel);
+	AddPanelString(tempstr, true);
+
+	sprintf(tempstr, "Experience: ");
+	PrintWithSeparator(tempstr + SDL_arraysize("Experience: ") - 1, player._pExperience);
+	AddPanelString(tempstr, true);
+
+	sprintf(tempstr, "Next Level: ");
+	PrintWithSeparator(tempstr + SDL_arraysize("Next Level: ") - 1, ExpLvlsTbl[charLevel]);
+	AddPanelString(tempstr, true);
 
-	FillRect(out, xPos, yPos, barWidth, barHeight, emptyBarColor);
-	FastDrawHorizLine(out, xPos - 1, yPos - 1, barWidth + 2, frameColor);
-	FastDrawHorizLine(out, xPos - 1, yPos + barHeight, barWidth + 2, frameColor);
-	FastDrawVertLine(out, xPos - 1, yPos - 1, barHeight + 2, frameColor);
-	FastDrawVertLine(out, xPos + barWidth, yPos - 1, barHeight + 2, frameColor);
-	for (int i = 1; i < numDividers; i++)
-		FastDrawVertLine(out, xPos - 1 + (barWidth * i / numDividers), yPos - dividerHeight + 3, barHeight, 245);
+	sprintf(PrintWithSeparator(tempstr, ExpLvlsTbl[charLevel] - player._pExperience), " to Level %d", charLevel + 1);
+	AddPanelString(tempstr, true);
 
-	FillRect(out, xPos, yPos + (space ? 1 : 0), visibleBar, barHeight - (space ? 2 : 0), barColor);
+	return true;

```

Comment:
I don't know why the diff felt the need to mix the xp bar drawing code into the xp info printing code. Diff tools can be like that sometimes.

I think the functions are short enough that they can be reviewed on their own, however, rather than a line by line comparison with the previous code.